### PR TITLE
 Apple attestation implementation

### PR DESCRIPTION
**Pull request by randomprogramming**
PR #87

### CODE REVIEW

**Comment by JamesCullum** [comment]
Please let me know once this is ready for review 👍

**Comment by randomprogramming** [comment]
@JamesCullum  Should be ready 👍

File: lib/attestations/apple.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	const certificatePublicKeyReexportedPem = await certificatePublicKey.toPem(
+		true
+	);
+	this.audit.journal.add("credentialPublicKeyPem");
+
+	if (credentialPublicKeyReexportedPem !== certificatePublicKeyReexportedPem) {
+		throw new Error("certificatePublicKey did not match credentialPublicKey");
+	}
+
+	// ToDo: Verify certificate chain?
```
Comment (JamesCullum) [inline lib/attestations/apple.js:96]: Can you elaborate why this is not needed for safe validation in the beginning? People probably assume that we do all cryptographic validations if we say that we are compatible.

File: lib/attestations/apple.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+	if (!arrayBufferEquals(expectedNonce, nonce)) {
+		throw new Error("nonce did not match expectedNonce");
+	}
+
+	// Step 5: Verify that the credential public key equals the Subject Public Key of credCert.
+	const credentialPublicKey = new PublicKey();
+	await credentialPublicKey.fromPem(
+		this.authnrData.get("credentialPublicKeyPem")
+	);
+	console.log(await credCert.getPublicKey());
```
[the remark this answers is not in the document]
Comment (Hexagon) [reply lib/attestations/apple.js:81]: My bad :)

File: lib/attestations/apple.js (hunk 1 above)
Comment (Hexagon) [reply lib/attestations/apple.js:96]: Wierd thing is that the spec says nothing about this

https://www.w3.org/TR/webauthn-2/#sctn-apple-anonymous-attestation

It almost seems like the RP is supposed to do this?

File: lib/attestations/apple.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hexagon) [reply lib/attestations/apple.js:81]: @randomprogramming Fixed in a PR to to this pr :)

File: lib/attestations/apple.js (hunk 1 above)
Comment (JamesCullum) [reply lib/attestations/apple.js:96]: Good point - I dont see a specific validation step for this either, so I'd say we're good by following the standard. Can we remove the comment then to remove confusion?

Comment (Hexagon) [reply lib/attestations/apple.js:96]: 👍 

Submitted a "third party" PR

https://github.com/randomprogramming/fido2-lib/pull/4/files

**Comment by randomprogramming** [comment]
@JamesCullum Should be all ready now? Sorry didn't get a notification for the PR.

**Comment by Hexagon** [comment]
@randomprogramming Bump version while we're waiting, not sure if its a patch or minor though?

**Comment by randomprogramming** [comment]
Seems like a minor to me, this is a pretty significant addition imo.

**Comment by JamesCullum** [comment]
Current code scanning fails because of a bad regex: 
`thing = [...].replace(/=*$/g, "");`

Details: https://codeql.github.com/codeql-query-help/javascript/js-polynomial-redos/

I understand that you're trying to remove all "=" at the end of the string - there is probably a reasonable assumption we can make about how many possible hits there are?!

**Comment by Hexagon** [comment]
This is unrelated to the pull request, but  0 to 2 occurences of = should be replaced at that point. Maybe @randomprogramming can fix to clear this alert?

**Comment by randomprogramming** [comment]
@JamesCullum It's running the tests on the build files, right? Would I need to bump the version again to make it build with the updated regex?

**Comment by Hexagon** [comment]
@randomprogramming no need to bump again 👍

**Comment by Hexagon** [comment]
The dists need to be updated with the same fix, but that's made by CI. Should be good to go now? @JamesCullum

**Comment by JamesCullum** [comment]
Great, thanks everyone! Merging now - let me know if anything goes wrong 👍

**Comment by randomprogramming** [comment]
Thanks @JamesCullum and thanks @Hexagon !

**Pull request merged**